src/core/thread.d

```diff
@@ -5018,6 +5089,80 @@ unittest
     group.joinAll();
 }
 
+// Try to detect if version CheckFiberMigration should be set, or if it is
+// set, make sure it behaves properly.  The issue is that thread local addr
+// may be cached by the compiler and that doesn't play well when Fibers
+// migrate between Threads.  This may only happen when optimization
+// enabled.  For that reason, should run this unittest with various
+// optimization levels.
+//
+// https://github.com/ldc-developers/ldc/issues/666
+unittest
+{
+    static int tls;
+
+    static yield_noinline()
+    {
+        import ldc.intrinsics;
+        pragma(LDC_never_inline);
+        Fiber.yield();
+    }
+    
+    auto f = new Fiber(
+    {
+        ++tls;                               // happens on main thread
+        yield_noinline();
+        ++tls;                               // happens on other thread,
+        // 1 if tls addr uncached, 2 if addr was cached
+    });
+
+    auto t = new Thread(
+    {
+        assert(tls == 0);
+
+        version( CheckFiberMigration )
+        {
+            try
+            {
+                f.call();
+                assert(false, "Should get FiberException when Fiber migrated");
+            }
+            catch (FiberException ex)
+            {
+            }
+
+            f.allowMigration();
+        }
+
+        f.call();
+        // tls may be 0 (wrong) or 1 (good) depending on thread local handling
+        // by compiler
+    });
+
+    assert(tls == 0);
+    f.call();
+    assert(tls == 1);
+
+    t.start();
+    t.join();
+
+    version( CheckFiberMigration )
+    {
+        assert(Fiber.migrationUnsafe);
+    }
+    else
+    {
+        assert(!Fiber.migrationUnsafe);
+        // If thread local addr not cached (correct behavior), then tls should
+        // still be 1.
+        assert(tls != 2,
+               "Not safe to migrate Fibers between Threads on your system. "
+               "Consider setting version CheckFiberMigration for this system "
+               "in thread.d");
+        // verify un-cached correct case
```

Comment:
ok, done - but it only got one additional build green.  There is more going on here.